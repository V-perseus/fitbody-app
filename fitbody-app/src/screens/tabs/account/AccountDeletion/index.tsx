import AccountDeletion from './AccountDeletion'

export default AccountDeletion
