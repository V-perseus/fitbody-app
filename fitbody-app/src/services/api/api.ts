import * as Sentry from '@sentry/react-native'
import deviceInfoModule from 'react-native-device-info'
import { Platform } from 'react-native'

// Services
import { setErrorMessage } from '../error'
import { displayLoadingModal, hideLoadingModal } from '../loading'
import NavigationService from '../NavigationService'
import globals from '../../config/globals'
import { updateUserIsRestricted } from '../../data/user'
import { storeRegistry } from '../../store/storeRegistry'

const API_BASE = globals.apiBase

type Methods = 'get' | 'post' | 'patch' | 'put' | 'delete'
type MethodsUppercase = 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE'
interface IDefaultHeaders {
  [key: string]: string
}
interface IParams {
  method: MethodsUppercase
  headers: IDefaultHeaders
  body?: FormData | string
}
export interface IAppErrorResponse {
  response: {
    data: {
      message: string
    }
  }
}
export function isAppErrorResponse(error: any): error is IAppErrorResponse {
  return error.response?.data?.message !== undefined
}

export const apiPublic = <T>(url: string, method: Methods = 'get', data: {} | null = {}, loader = true, headers = {}): Promise<T> => {
  // console.log('api',`${API_BASE}${url}`);
  if (loader) {
    displayLoadingModal()
  }

  let token = storeRegistry.getState().services.session.apiToken
  let defaultHeaders: IDefaultHeaders = {
    'Content-Type': 'application/json',
    Accept: 'application/json',
    'X-App-Version': deviceInfoModule.getVersion(),
    'X-App-Os': Platform.OS,
  }
  if (token) {
    defaultHeaders.Authorization = 'Bearer ' + token
  }

  let reqHeaders = {}
  Object.assign(reqHeaders, defaultHeaders)
  let params: IParams = {
    method: method.toUpperCase() as MethodsUppercase,
    headers: reqHeaders,
  }

  if ((method.toUpperCase() === 'POST' || method.toUpperCase() === 'PUT' || method.toUpperCase() === 'PATCH') && data) {
    if (data instanceof FormData) {
      params.body = data
    } else {
      params.body = JSON.stringify(data)
    }
  }

  let req = new Request(API_BASE + url, params)

  return fetch(req)
    .then((response) => {
      if (response.ok) {
        if (loader) {
          hideLoadingModal()
        }
        let responseText = response.text()
        return responseText
          .then(function (text) {
            return text ? Promise.resolve(JSON.parse(text)) : Promise.resolve({})
          })
          .catch(function (error) {
            return Promise.reject(error)
          })
      }
      if (loader) {
        hideLoadingModal()
      }
      return Promise.reject(response)
    })
    .catch((error: Response) => {
      let message: string | Error = ''
      if (error && error instanceof Error) {
        try {
          Sentry.captureException(new Error(error.message || JSON.stringify(error)))
        } catch {}
        message = error
      } else if (error && error.json) {
        error.json().then((errorJson) => {
          message = errorJson.message
          if (message && typeof message === 'string' && message.toLowerCase().includes('please purchase')) {
            updateUserIsRestricted(true)
            NavigationService.sendToPaywall()
          } else if (!url.includes('/v4/events')) {
            setErrorMessage({ error: message })
          }
        })
      }
      hideLoadingModal()
      return Promise.reject(message)
    })
}

/**
 * Sign a user in
 */
export const apiAuth = <T>(url: string, method: Methods = 'get', data: null | {} = {}, loader = true, headers = {}): Promise<T> => {
  return apiPublic<T>(url, method, data, loader, headers)
}
