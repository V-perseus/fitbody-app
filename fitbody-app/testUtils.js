import React from 'react'
import { render as rtlRender } from '@testing-library/react-native'
import { configureStore } from '@reduxjs/toolkit'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import { Provider } from 'react-redux'

// Import your own reducer
import { appReducer } from './src/store'
import { ModalStackScreens } from './src/config/routes/loggedInStack'

// custom render method override
function render(ui, { preloadedState, store = configureStore({ reducer: appReducer, preloadedState }), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react-native'
// override render method
export { render }

// custom navigation mock
const Stack = createStackNavigator()
export const MockedNavigator = ({
  component,
  params = {},
  preloadedState,
  store = configureStore({ reducer: appReducer, preloadedState }),
}) => {
  return (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name="MockedScreen" component={component} initialParams={params} />
          <Stack.Screen name="Modals" component={ModalStackScreens} />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  )
}
