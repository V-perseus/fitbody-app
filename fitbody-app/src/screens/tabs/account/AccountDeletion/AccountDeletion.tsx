import React, { useLayoutEffect } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

// Types
import { AccountDeletionScreenUseNavigationProp } from '../../../../config/routes/routeTypes'

// Components
import { HeaderButton } from '../../../../components/Buttons/HeaderButton'
import FocusAwareStatusBar from '../../../../shared/FocusAwareStatusBar'
import { ButtonSquare } from '../../../../components/Buttons/ButtonSquare'

// Styles
import globals from '../../../../config/globals'

// Data
import api from '../../../../services/api'
import { setErrorMessage } from '../../../../services/error'
import { checkAndRedirect } from '../../../../config/utilities'

interface IAccountDeletionProps {
  navigation: AccountDeletionScreenUseNavigationProp
}
const AccountDeletion: React.FC<IAccountDeletionProps> = ({ navigation }) => {
  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitle: () => <Text style={styles.headerTitle}>DATA DELETION REQUEST</Text>,
      headerLeft: () => <HeaderButton onPress={() => navigation.goBack()} />,
    })
  })

  async function handleDeleteRequest() {
    try {
      const resp = await api.users.requestAccountDelete()
      if (resp.sent) {
        navigation.navigate('Modals', {
          screen: 'ConfirmationDialog',
          params: {
            showCloseButton: true,
            yesLabel: 'GOT IT!',
            hideNoButton: true,
            iconType: 'check',
            yesHandler: handleBack,
            title: 'Account Deletion Request Submitted!',
            body: 'We have received your request for your account and data to be deleted. Further action may be required, so please be on the lookout for a follow-up email from us in 1-2 business days.',
          },
        })
      } else {
        setErrorMessage({ error: 'There was an error submitting your request. Please try again later.' })
      }
    } catch (error) {
      // Will get handled by global error handler
    }
  }

  function handleBack() {
    navigation.goBack()
  }

  return (
    <SafeAreaView style={styles.container}>
      <FocusAwareStatusBar barStyle="dark-content" />
      <View style={styles.view}>
        <Text style={styles.leaderText}>
          This action will delete your workout history and personal data inside the app.{' '}
          <Text style={{ fontWeight: 'bold' }}>This action will not cancel your membership. </Text>
          In order to cancel your membership, please see our FAQ or email us at{' '}
          <Text style={{ fontWeight: 'bold' }} onPress={() => checkAndRedirect('nutrition@fitbodyapp.com')}>
            hello@fitbodyapp.com.
          </Text>
        </Text>
        <Text style={[styles.leaderText, { marginBottom: 28 }]}>
          If you would like to proceed with deleting your personal data, confirm below. Please note that this action cannot be undone and
          you will immediately lose access to the Fit Body App and all your workout progress and history.
        </Text>
        <ButtonSquare
          onPress={handleDeleteRequest}
          text="YES, DELETE MY PERSONAL DATA"
          style={styles.deleteButton}
          textStyle={styles.deleteButtonText}
        />
        <ButtonSquare onPress={handleBack} text="NO THANKS" style={styles.cancelButton} textStyle={styles.cancelButtonText} />
      </View>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container: {
    ...globals.styles.container,
    alignItems: 'center',
    backgroundColor: globals.styles.colors.colorWhite,
  },
  view: {
    ...globals.styles.view,
    flex: 1,
    backgroundColor: globals.styles.colors.colorWhite,
    alignItems: 'center',
    marginHorizontal: 25,
  },
  headerTitle: {
    ...globals.header.headerTitleStyle,
  },
  leaderText: {
    fontSize: 14,
    textAlign: 'center',
    fontFamily: globals.styles.fonts.primary.style.fontFamily,
    color: globals.styles.colors.colorBlack,
    marginVertical: 12,
  },
  deleteButton: {
    borderWidth: 1,
    borderColor: globals.styles.colors.colorBlack,
    width: globals.window.width - 50,
    height: 56,
    marginBottom: 24,
  },
  deleteButtonText: {
    color: globals.styles.colors.colorBlack,
    fontFamily: globals.styles.fonts.primary.boldStyle.fontFamily,
    fontSize: 14,
  },
  cancelButton: {
    backgroundColor: globals.styles.colors.colorPink,
    width: globals.window.width - 50,
    height: 56,
  },
  cancelButtonText: {
    fontFamily: globals.styles.fonts.primary.boldStyle.fontFamily,
    fontSize: 14,
  },
})

export default AccountDeletion
