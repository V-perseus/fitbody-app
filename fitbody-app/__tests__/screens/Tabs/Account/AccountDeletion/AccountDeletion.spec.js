import { MockedNavigator, render, cleanup, fireEvent } from '../../../../../testUtils'

import AccountDeletion from '../../../../../src/screens/tabs/account/AccountDeletion'

import api from '../../../../../src/services/api'

jest.mock('../../../../../src/services/api')

describe('Account Deletion screen', () => {
  let component = null
  beforeEach(() => {
    component = MockedNavigator({ component: AccountDeletion })
  })
  afterEach(() => {
    cleanup()
  })

  it('renders snapshot correctly', async () => {
    const tree = render(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render header buttons', async () => {
    const { findByText } = render(component)
    await findByText('DATA DELETION REQUEST')
  })

  it('should render correct buttons and body text', async () => {
    const { findByText } = render(component)
    await findByText(
      /This action will delete your workout history and personal data inside the app. This action will not cancel your membership. In order to cancel your membership, please see our FAQ or email us at/i,
    )
    await findByText(
      /If you would like to proceed with deleting your personal data, confirm below. Please note that this action cannot be undone and you will immediately lose access to the Fit Body App and all your workout progress and history./i,
    )
    await findByText(/YES, DELETE MY PERSONAL DATA/gi)
    await findByText(/NO THANKS/gi)
  })

  it('should submit request and show confirmation', async () => {
    api.users.requestAccountDelete.mockResolvedValueOnce({ sent: true, message: 'Thanks for playing' })
    const { findByText } = render(component)
    const delBtn = await findByText(/YES, DELETE MY PERSONAL DATA/gi)
    fireEvent.press(delBtn)
    await findByText(/Account Deletion Request Submitted/gi)
    await findByText(
      /We have received your request for your account and data to be deleted. Further action may be required, so please be on the lookout for a follow-up email from us in 1-2 business days./i,
    )
  })
})
