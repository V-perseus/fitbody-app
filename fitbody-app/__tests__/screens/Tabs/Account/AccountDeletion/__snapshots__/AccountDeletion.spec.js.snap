// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Deletion screen renders snapshot correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  hitSlop={16}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "opacity": 1,
                      },
                      Object {
                        "alignItems": "center",
                        "borderRadius": 30,
                        "display": "flex",
                        "height": 42,
                        "justifyContent": "center",
                        "marginHorizontal": 7,
                        "width": 42,
                      },
                      undefined,
                    ]
                  }
                >
                  <arrow-left.svg
                    color="#333333"
                    testID="back_arrow"
                  />
                </View>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 590,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontFamily": "OpenSans-Semibold",
                      "fontSize": 16,
                      "textAlign": "center",
                    }
                  }
                >
                  DATA DELETION REQUEST
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RNCSafeAreaView
                        edges={
                          Array [
                            "bottom",
                            "left",
                            "right",
                            "top",
                          ]
                        }
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "#ffffff",
                            "flex": 1,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "#ffffff",
                              "flex": 1,
                              "flexDirection": "column",
                              "marginHorizontal": 25,
                              "marginTop": 60,
                            }
                          }
                        >
                          <Text
                            style={
                              Object {
                                "color": "#333333",
                                "fontFamily": "OpenSans",
                                "fontSize": 14,
                                "marginVertical": 12,
                                "textAlign": "center",
                              }
                            }
                          >
                            This action will delete your workout history and personal data inside the app.
                             
                            <Text
                              style={
                                Object {
                                  "fontWeight": "bold",
                                }
                              }
                            >
                              This action will not cancel your membership. 
                            </Text>
                            In order to cancel your membership, please see our FAQ or email us at
                             
                            <Text
                              onPress={[Function]}
                              style={
                                Object {
                                  "fontWeight": "bold",
                                }
                              }
                            >
                              hello@fitbodyapp.com.
                            </Text>
                          </Text>
                          <Text
                            style={
                              Array [
                                Object {
                                  "color": "#333333",
                                  "fontFamily": "OpenSans",
                                  "fontSize": 14,
                                  "marginVertical": 12,
                                  "textAlign": "center",
                                },
                                Object {
                                  "marginBottom": 28,
                                },
                              ]
                            }
                          >
                            If you would like to proceed with deleting your personal data, confirm below. Please note that this action cannot be undone and you will immediately lose access to the Fit Body App and all your workout progress and history.
                          </Text>
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "opacity": 1,
                                },
                                Object {
                                  "alignItems": "center",
                                  "borderRadius": 8,
                                  "height": 56,
                                  "justifyContent": "center",
                                  "width": 114,
                                },
                                Object {
                                  "borderColor": "#333333",
                                  "borderWidth": 1,
                                  "height": 56,
                                  "marginBottom": 24,
                                  "width": 700,
                                },
                              ]
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#ffffff",
                                    "fontSize": 16,
                                  },
                                  Object {
                                    "color": "#333333",
                                    "fontFamily": "OpenSans-Bold",
                                    "fontSize": 14,
                                  },
                                ]
                              }
                            >
                              YES, DELETE MY PERSONAL DATA
                            </Text>
                          </View>
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "opacity": 1,
                                },
                                Object {
                                  "alignItems": "center",
                                  "borderRadius": 8,
                                  "height": 56,
                                  "justifyContent": "center",
                                  "width": 114,
                                },
                                Object {
                                  "backgroundColor": "#fa8dd5",
                                  "height": 56,
                                  "width": 700,
                                },
                              ]
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#ffffff",
                                    "fontSize": 16,
                                  },
                                  Object {
                                    "fontFamily": "OpenSans-Bold",
                                    "fontSize": 14,
                                  },
                                ]
                              }
                            >
                              NO THANKS
                            </Text>
                          </View>
                        </View>
                      </RNCSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
